Cython/Compiler/ExprNodes.py

```diff
@@ -5057,6 +5063,30 @@ def is_simple(self):
 
         return self.result_in_temp()
 
+    def call_funcstate_allocate_temp(self, code, type):
+        from .UtilNodes import ResultRefNode
+        # make the type distinct from the "normal" memoryview type
+
+        type = type.make_dont_clear_temps_type()
+        # unwrap result-refs to deal with the optimized
+        #  for x in memview:
+        # direct iteration. (It'd still work without this but
+        # it wouldn't preserve the "same temp for same entry" which
+        # should make the speed-up more reliable)
+        base = self.base
+        if isinstance(base, ResultRefNode):
+            base = base.expression
+        if isinstance(base, NoneCheckNode):
+            base = base.arg
```

Comment:
Isn't there already an `unpack()` function somewhere? I strongly suggest not distributing the knowledge about "which nodes to unwrap" (and in which order) all over the place. If we ever add new wrapping nodes or change the way the nodes are wrapped, we'll almost certainly forget to update at least some places where they were previously unwrapped.

We could have a generic `ExprNode.unpacked_node()` method that returns `self` unless specially overridden by utility nodes. That would be nice because it would also work recursively, with multiply wrapped nodes. But maybe that's not something for right now. Could be done as part of the 3.1 cleanup.